Add abort API in gloo ProcessGroup Send/Recv Work

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29852 Add abort API in gloo ProcessGroup Send/Recv Work**

- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

xush6528 committed:
Add abort API in gloo ProcessGroup Send/Recv Work

- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -33,13 +33,20 @@ std::vector<at::Tensor> ProcessGroup::Work::result() const {
+  return !aborted_;
+}
+
+void ProcessGroup::Work::abort() {
```
Just to confirm, we can only abort for `ProcessGroupGloo`, correct? Adding this here would make this default implementation available to `ProcessGroupNCCL` and `ProcessGroupMPI`, and since it doesn't do anything other than flip a boolean flag, I wonder if we should add a comment mentioning this?

xush6528 replied:
Correct. I only added this capability to ProcessGroupGloo to quickly unblock us. 

For  NCCL Work and MPI AsyncWork, I always returns true for wait().

Only MPI `WorkMPI` is using ProceessGroup::Work base implementation.

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -363,6 +364,12 @@ void ProcessGroupGloo::SendWork::wait() {
+
+void ProcessGroupGloo::SendWork::abort() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  buffer_->abortWaitSend();
```
should we also set `aborted_ = true` here for completeness?

xush6528 replied:
I don't feel it's the right thing to do. 
I think If a sub class doesn't use a field in the base class, just don't bother to use and modify it.

xush6528 committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -33,13 +33,20 @@ std::vector<at::Tensor> ProcessGroup::Work::result() const {
     std::rethrow_exception(exception_);
   }
   synchronize();
+  return !aborted_;
```
Should we change these default implementations given that we're only aiming to make this modification specific to `ProcessGroupGloo`?

xush6528 replied:
Good point. We may not be able to test it exhaustedly. Let me just raise if a user calls the ProcessGroup::abort just like NCCL and MPI.

xush6528 committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

xush6528 commented:
Revert changes to `ProcessGroup::Work::wait()` to minimize this PR.

rohan-varma commented:
Resubmitted at https://github.com/pytorch/pytorch/pull/29928.

xush6528 commented:
Moved to #29928.
